src/test/clojure/xtdb/trie_test.clj

```diff
@@ -173,23 +173,89 @@
   ([page metadata-matches?] (->MockMergePlanPage page metadata-matches? (TemporalBounds.)))
   ([page min-valid-from max-valid-to] (->mock-merge-plan-page page true min-valid-from max-valid-to))
   ([page metadata-matches? min-valid-from max-valid-to]
-   (->MockMergePlanPage page metadata-matches? (tu/->min-max-page-bounds min-valid-from max-valid-to))))
+   (->MockMergePlanPage page metadata-matches? (tu/->min-max-page-bounds min-valid-from max-valid-to)))
+  ([page metadata-matches? min-vf max-vt min-sf]
+   (->mock-merge-plan-page page metadata-matches? min-vf max-vt min-sf Long/MAX_VALUE))
+  ([page metadata-matches? min-vf max-vt min-sf max-st]
+   (->MockMergePlanPage page metadata-matches? (tu/->min-max-page-bounds min-vf max-vt min-sf max-st)))
+  ([page metadata-matches? min-vf max-vt min-sf max-st max-vf]
+   (->MockMergePlanPage page metadata-matches? (tu/->min-max-page-bounds min-vf max-vt min-sf max-st max-vf))))
 
-(deftest test-to-merge-task
-  (t/is (= [1 2] (->> (trie/->merge-task [(->mock-merge-plan-page 0 false)
-                                          (->mock-merge-plan-page 1 true)
-                                          (->mock-merge-plan-page 2 false)])
-                      (map :page)))
-        "earlier pages don't need to be taken if metadata doesn't match")
-
-  (t/is (= [0 1 2] (->> (trie/->merge-task [(->mock-merge-plan-page 0 true)
-                                            (->mock-merge-plan-page 1 false)
-                                            (->mock-merge-plan-page 2 false)])
-                        (map :page)))
-        "all pages pages need to be taken if the earliest matches"))
+(def ^:private inst->micros (comp time/instant->micros time/->instant))
 
+(deftest earilier-recency-buckets-can-effect-splitting-in-later-buckets-4097
+  (let [[system-from1 system-from2] (mapv inst->micros [#inst "2020-01-01" #inst "2020-01-03"])
+        [valid-from1 valid-from2] (mapv inst->micros [#inst "2020-01-01" #inst "2020-01-03"])
+        [valid-to1 valid-to2] (mapv inst->micros [#inst "2021-01-05" #inst "2020-01-04"])]
+    (t/is (= [1 0] (->> (trie/->merge-task
+                         [(->mock-merge-plan-page 0 false valid-from1 valid-to1 system-from1)
+                          (->mock-merge-plan-page 1 true  valid-from2 valid-to2 system-from2)])
+                        (map :page)))
+          "earlier pages that contain data with later system time need to be taken")))
 
-(def ^:private inst->micros (comp time/instant->micros time/->instant))
+(deftest test-to-merge-task
+  (let [insts (tu/->instants :day)
```

Comment:
use numbers rather than instants, save the round-trip, then we can just reason about raw nums